src/django_otp/models.py

```diff
@@ -186,6 +190,49 @@ def verify_token(self, token):
         return False
 
 
+class SideChannelDevice(Device):
+    """
+    Abstract base model for a side-channel :term:`device` attached to a user.
+    This model implements token generation, verification and expiration, so
+    the concrete devices only have to implement delivery.
+    """
+    token = models.CharField(max_length=16, blank=True, null=True)
+    valid_until = models.DateTimeField(
+        default=timezone.now,
+        help_text="The timestamp of the moment of expiry of the saved token."
+    )
+
+    class Meta:
+        abstract = True
+
+    def generate_token(self, length=6, valid_secs=300):
+        """
+        Generates a token of the specified length, then sets it on the model
+        and sets the expiration of the token on the model.
+
+        The concrete device should save the object after calling this method.
+
+        :param int length: Length of the generated token.
+        :param int valid_secs: Amount of seconds the token should be valid.
+        """
+        self.token = random_number_token(length)
+        self.valid_until = timezone.now() + timedelta(seconds=valid_secs)
+
+    def verify_token(self, token):
+        """
+        Verifies a token by content and expiry. When valid, the expiry time
+        is updated to the current time, to counter token reuse attacks.
+
+        :param string token: The OTP token provided by the user.
+        :rtype: bool
+        """
+        if token == self.token and self.valid_until >= timezone.now():
+            self.valid_until = timezone.now()
```

Comment:
Strictly speaking, the previous conditional will still succeed for very short time after this statement. Maybe use a `>` test to remove the race condition?

Reply:
Thank you, I missed that. It's fixed now.